kolibri/core/tasks/worker.py

```diff
@@ -81,16 +86,23 @@ def start_workers(self):
 
     def handle_finished_future(self, future):
         # get back the job assigned to the future
-        job = self.job_future_mapping[future]
-
-        # Clean up tracking of this job and its future
-        del self.job_future_mapping[future]
-        del self.future_job_mapping[job.job_id]
-
-        try:
-            future.result()
-        except CancelledError:
-            self.storage.mark_job_as_canceled(job.job_id)
+        job = None 
+
+        # Acquire locks before accessing dictionaries
+        with self.job_future_mapping_lock:
+            with self.future_job_mapping_lock:
+                if future in self.job_future_mapping:
```

Comment:
More so than the locks, it seems like this line is the one doing most of the work here?

Your test above is not obeying the locks at all, so this existence check is preventing errors from occurring, in which case, this may just work without the locks at all.